### PR TITLE
Panic on invalid usages of `MaybeUninit::uninit().assume_init()`

### PR DESCRIPTION
**Pull request by Noratrieb**
This MIR transform inserts the same validity checks from `mem::{uninitialized,zeroed}` to `MaybeUninit::{uninit,zeroed}().assume_init()`.

We have been panicking in `mem::uninit` on invalid values for quite some time now, and it has helped to get people off the unsound API and towards using `MaybeUninit<T>`.

While correct usage of `MaybeUninit<T>` is clearly documented, some people still use it incorrectly and simply replaced their wrong `mem::uninit` usage with `MaybeUninit::uninit().assume_init()`. This is not any more correct than the old version, and we should still emit panics in these cases. As this can't be done in the library only, we need this MIR pass to insert the calls.

For now, it only detects direct usages of `MaybeUninit::uninit().assume_init()` but it could be extended in the future to do more advanced dataflow analysis.

### CODE REVIEW

**Comment by 5225225** [comment]
Also gonna leave a linkback to https://github.com/rust-lang/rust/issues/66151 which this PR is relevant to... Kinda!

File: library/core/src/intrinsics.rs [hunk 1]
```diff
@@ -2714,3 +2714,30 @@ where
 {
     called_in_const.call_once(arg)
 }
+
+// Wrappers around the init assertion intrinsics. Calls to these
+// are inserted int he check_maybe_uninit mir pass
```
Comment (reitermarkus) [inline library/core/src/intrinsics.rs:2719]:
```suggestion
// are inserted in the check_maybe_uninit mir pass
```

**Comment by 5225225** [comment]
Heads up, with `zstd-seekable 0.1.7`, `RUSTFLAGS="-Zstrict-init-checks" cargo test` just hangs for me (in *test*, not in build). Not sure if you can reproduce. No idea what it's doing.

**Comment by michaelwoerister** [comment]
r? rust-lang/mir

**Comment by 5225225** [comment]
> Heads up, with `zstd-seekable 0.1.7`, `RUSTFLAGS="-Zstrict-init-checks" cargo test` just hangs for me (in _test_, not in build). Not sure if you can reproduce. No idea what it's doing.

Did some investigation with @Nilstrieb and it seems to be their problem. We're inserting panics in a place they're not panic safe. They deadlock. Happens on stable if you insert a panic manually.

File: compiler/rustc_mir_transform/src/lib.rs [hunk 2]
```diff
@@ -242,6 +243,8 @@ fn mir_const<'tcx>(
             &Lint(function_item_references::FunctionItemReferences),
             // What we need to do constant evaluation.
             &simplify::SimplifyCfg::new("initial"),
+            // FIXME IN REVIEW: This is probably not the correct place for this.
+            &check_maybe_uninit::CheckMaybeUninit,
```
Comment (JakobDegen) [inline compiler/rustc_mir_transform/src/lib.rs:247]: I don't think there's anything wrong with putting it here. It doesn't really matter. I guess we want this to block optimizations, so putting it before that is the only important thing.

**Comment by Nilstrieb** [comment]
This will take a little time as I don't have that much time right now, so @rustbot author

**Comment by RalfJung** [comment]
Honestly I am not sure this is the right move. With a panic like this, the chances are pretty high that they will hit people that can do absolutely nothing about them -- because the panic is caused by a dependency. And having to implement this via a MIR pass sounds like the kind of hack that will be quite annoying to maintain.

I think I'd be more in favor of using a lint for this. And note that the `invalid_values` lint already detects `MaybeUninit::uninit().assume_init()`. Also IIRC @5225225 had some great ideas for how to make `assume_init` panic on invalid bit patterns, which can be enabled by a `-Z` flag.

**Comment by 5225225** [comment]
One advantage the panics have over a lint is the lint needs to work pre-monomorphisation, so we can't lint for making an uninit `T`, because that *might* be fine.

And the lint warns about dead code, which is arguably a benefit (don't need to actually test the code, you get immediate feedback).

The panics run where we *know* the type layout concretely, so we can accurately panic even in cases of zeroed enums where the variant with discriminant zero doesn't allow being left zeroed. That would be very hard, if not impossible, to use a lint for.

That's not to say our lints couldn't be improved, but I think the panics serve a useful purpose by saying you *did do* a UB. Both are useful in different circumstances, which is why I think we should have both.

File: compiler/rustc_mir_transform/src/check_maybe_uninit.rs [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+            let Some(assume_init_bb) = assume_init_bb else {
+                continue;
+            };
+
+            let Some((assume_init_operand, assume_init_call_span)) = is_block_just_assume_init(tcx, &body.basic_blocks[*assume_init_bb]) else {
+                continue;
+            };
+
+            let Operand::Move(assume_init_place) = assume_init_operand else {
```
[the remark this answers is not in the document]
Comment (Nilstrieb) [reply compiler/rustc_mir_transform/src/check_maybe_uninit.rs:50]: I can't figure out how to make it generate a `Copy`.

**Comment by cuviper** [comment]
> With a panic like this, the chances are pretty high that they will hit people that can do absolutely nothing about them -- because the panic is caused by a dependency.

If you replace "panic" in your statement with "UB", is the meaning much different? The dependency is behaving badly, which is hard for its users, but the panic at least makes it deterministic.

**Comment by 5225225** [comment]
Also, it's worth thinking about if we're trying to prevent just LLVM UB here, or rust UB? I agree that being minimal in our checks for `mem::uninitialized` is useful because that function is what Old Code does, which is why we added the 0x01 filling, but can/should we be stricter and go "we will panic on *rust* UB" for MaybeUninit, since we sorted out our messaging (mostly) for what uninit means?

(We did say "uhhhh uninit ints may be fine, they may not, be careful regardless", which has since been changed to "they're UB.")

If it's rust UB, then strict-init-checks is the correct thing to do here.

If it's LLVM UB, then we need to still panic on uninit `bool`, but not behind arrays. We wouldn't panic on uninit `u8` until we start adding `noundef` to ints (which it would be a shame if we *couldn't* do).

**Comment by 5225225** [comment]
Though I do think the general concept of "change the code during compilation to detect more UB" is something we should support. Other things we could do along the same lines, that would have the same downside of making the resulting MIR not match the input code, include "check type validity on pointer reads" or "check pointer alignment on read".

What I'm thinking of is having a set of sanitizers that are on by default, with the default set depending on profile, and letting you disable them if you wanted.

So something like `RUSTFLAGS="-Cmir-sanitizer=uninit_assume_init,pointer_align" cargo build` (or a similar thing in your Cargo.toml `[profile.dev]`. If it's just A Done Thing that "yeah, your code will be modified sometimes to insert UB checks, use this flag to disable that/only enable *some* of them", I don't think it would be too bad?

Also, could use a fuel mechanism like optimizations do (maybe the same fuel? these aren't *required* transformations, they're just not optimizations) in order to let you bisect bad transformations.

Basically, I think these shouldn't be ad-hoc changes that are impossible to control, but I think we *should* have the infrastructure to do changes like this, since they're useful.

**Comment by Nilstrieb** [comment]
Yeah, this pass could be seen as a very simple MaybeUninit sanitizer. But I agree that this is quite some complexity without precedent, and I am not sure whether it's worth it either. I think it's a good idea, which is why I wrote it in the first place, but in the end the choice is up to wg-mir-opt, and maybe going with more general MIR sanitizers like 522 mentioned is a better idea.

**Comment by oli-obk** [comment]
If we're talking about sanitizers, we could also change `MaybeUninit` to have an additional field behind `#[cfg(debug_assertions)]` that is used to track initialization and panics if you haven't used any of the `&mut self` methods before asserting intializedness.

That has zero impact on release mode, and doesn't require much maintenance

**Comment by Nilstrieb** [comment]
This would break, we guarantee it to be transparent over `T`. If I read a `T` in memory as a `MaybeUninit<T>`, which is allowed as it's transparent, it would do an out of bounds read of the init flag.

**Comment by cuviper** [comment]
Yeah, `MaybeUninit` has several slice methods that need matching layout.

**Comment by RalfJung** [comment]
> Though I do think the general concept of "change the code during compilation to detect more UB" is something we should support. Other things we could do along the same lines, that would have the same downside of making the resulting MIR not match the input code, include "check type validity on pointer reads" or "check pointer alignment on read".

Those are very different. We can compile each raw pointer read to a checked operation. That is a correct compilation scheme, and perfectly compositional -- the behavior of a large expression is entirely determined by its subexpressions.

What I am objecting to is having the behavior of `MaybeUninit::uninit().assume_init()` be something other than first calling `MaybeUninit::uninit()` and then calling `assume_init()` on the result.

**Comment by tavianator** [comment]
> If we're talking about sanitizers, we could also change `MaybeUninit` to have an additional field behind `#[cfg(debug_assertions)]` that is used to track initialization and panics if you haven't used any of the `&mut self` methods before asserting intializedness.
> 
> That has zero impact on release mode, and doesn't require much maintenance

The C sanitizers use "shadow memory" for this, where each stack/heap allocation has additional metadata at a predictable address that can track initialization status etc.  It would be interesting to have that infrastructure for MIR somehow.

**Comment by RalfJung** [comment]
This PR is also [being discussed on Zulip now](https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang/topic/Panics.20in.20MaybeUninit.3A.3Aassume_init). Quoting some of the things I said there:

"I'd much rather see effort spent towards having a version of assume_init that checks the invariants it can check, independent of what happened to the MaybeUninit before this method was called."
"[This] PR only covers the most extreme least-likely-to-surprise-people case so I dont think it makes enough of a dent in UB prevention to justify its cost. It would be different if it would be able to detect common subtle cases of UB, but it doesn't."

**Comment by Nilstrieb** [comment]
I'm waiting for whether we want to go forward with this before I'm cleaning this up further

**Comment by JohnCSimon** [comment]
ping from triage:
@Nilstrieb


> I'm waiting for whether we want to go forward with this before I'm cleaning this up further

This was over a month ago.
Can you please post your status on this PR?

FYI: when a PR is ready for review, send a message containing 
`@rustbot ready` to switch to `S-waiting-on-review` so the PR is in the reviewer's backlog.

**Comment by RalfJung** [comment]
This needs a decision by someone, but I don't know who? I guess it'd be the lang team for whether we want such a panic, and the compiler team for whether it is worth the maintenance burden?

**Comment by RalfJung** [comment]
Dear @rust-lang/compiler @rust-lang/lang,

this PR suggests to make `MaybeUninit::uninit().assume_init()` panic when it is called on a type that does not permit being uninitialized, similar to what we already do for `mem::uninitialized()`. The way this is implemented is via a MIR transform that detects this pattern (`MaybeUninit::assume_init` called on the return value of `MaybeUninit::uninit`), check the type, and rewrite it into a panic if applicable.

I am bringing this up for nomination because this is the first proposal for a MIR transform whose goal it is to improve run-time UB detection. So far, we have MIR transforms for optimizations, and we have assertions in the standard library for run-time UB detection. So I am asking:
- Compiler team, are you okay with having such MIR transforms around whose purpose is not optimization but UB detection?
- Lang team, do you think using a MIR transform for UB detection is a sensible thing to do?

I personally am concerned about the surprise that might be cause by these rewrites. Basically this means that the behavior of `MaybeUninit::uninit().assume_init()` is no longer determined by first calling `uninit` and then `assume_init`, there is now a non-compositional rewrite being applied that will *sometimes* change program behavior. (Naturally, the transform cannot catch all cases of `MaybeUninit::assume_init` being called on the return value of `MaybeUninit::uninit`.) I imagine if we land this, it would not be the last MIR transform of that kind.

On the other hand, it does help detect some UB, after all.

**Comment by scottmcm** [comment]
> We're inserting panics in a place they're not panic safe.

Would it make sense to have these be aborts instead?

**Comment by RalfJung** [comment]
These days we have `panic_str_nounwind` so yeah that should be possible (and we probably will [switch the existing panics to that](https://github.com/rust-lang/rust/issues/104451#issuecomment-1315767039), too).

**Comment by apiraino** [comment]
visiting during T-compiler meeting on [Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202022-11-17/near/310639666): letting T-lang have a say on this first.

@rustbot label -I-compiler-nominated

**Comment by RalfJung** [comment]
https://github.com/rust-lang/rust/pull/98112 is doing something fairly similar in spirit, so they should probably be discussed together.

**Comment by tmandry** [comment]
Discussed in this week's @rust-lang/lang meeting ([notes](https://hackmd.io/NtN56hozSSSlWGZJI82sHw?both#%E2%80%9CPanic-on-invalid-usages-of-MaybeUninituninitassume_init%E2%80%9D-rust100423)). There are some points that we would like to see addressed:

**Documentation.** It should be as clear as possible how to take action on an abort caused by one of these checks. Documentation should be easy to find. When code that was perceived as "working" breaks, maintainers are less willing to put in effort to understand what is going on, and will often reach for the quickest fix available in the moment.

**Debuggability.** As previously discussed, we should make sure it is possible to understand _how_ this panic/abort is happening. If looking at the library source for `assume_init`, for example, there should be some kind of breadcrumb that indicates the compiler can introduce one. This could take the form of an intrinsic with a comment, or even just a comment. The docs for the function should also be updated to talk about the possibility.

**Impact.** We want to understand the level of impact on users, both library authors and consumers, and map out what options they will have if their program aborts (especially if it's in a library they don't control). We may want to add mitigation options, like a compiler flag that disables these checks.

**Best effort.** We should communicate broadly and clearly that these checks are best-effort. We don't want users to rely on them as guarantees to prevent classes of UB, unless we can promise that and stick to that promise.

It's possible some of these require further discussion. Feel free to do that on Zulip #t-lang for smaller questions or [propose a design meeting](https://github.com/rust-lang/lang-team/issues/new?assignees=&labels=meeting-proposal%2C+T-lang&template=meeting-proposal.md&title=%28My+meeting+proposal%29) for longer discussion. If you feel comfortable addressing them in this thread, go ahead and do so and re-nominate the PR.

@rustbot label: -I-lang-nominated

Finally, it was noted that the general concept of runtime UB detection is one that would benefit from broader discussion and communication. In particular, we would be receptive to an RFC that discusses an overall approach to runtime UB detection and touches on these questions with future expansions in mind, like options for users to add more expensive runtime checks to their programs. We don't necessarily consider that a blocker for this PR, but think that it would be a useful direction for Rust's tooling to evolve in _and_ that it would be useful to do some initial thinking about how this PR fits into a broader picture.

File: library/core/src/intrinsics.rs [hunk 4]
```diff
@@ -2513,3 +2513,42 @@ where
...
+    feature = "const_maybe_uninit_checks_internals",
+    since = "CURRENT_RUSTC_VERSION"
+)]
+#[cfg_attr(not(bootstrap), lang = "assert_zero_valid")]
+#[track_caller]
+#[cfg_attr(not(bootstrap), allow(unused_unsafe))]
+pub const fn assert_zero_valid_wrapper<T>() {
+    // SAFETY: Always safe to call
+    unsafe { assert_zero_valid::<T>() };
+}
```
Comment (RalfJung) [inline library/core/src/intrinsics.rs:2536]: On the implementation side, wouldn't it be easier to directly insert calls to the intrinsics? Right now there is a library roundtrip and I cannot see a good reason for that.

Comment (Nilstrieb) [reply library/core/src/intrinsics.rs:2536]: It needs to be a lang item to insert a call to it. When I tried making the intrinsic a lang item, I had a very bad time.

Comment (RalfJung) [reply library/core/src/intrinsics.rs:2536]: Ah `Call` wants a DefId... we only care about its name when it is an intrinsic so it's a bit silly, but it's what it is I guess.

**Comment by Nilstrieb** [comment]
I would prefer to block this on #100342 so that we fire out an FCW for obviously offending code first.
@rustbot label +S-blocked -S-waiting-on-author

**Comment by Dylan-DPC** [comment]
Closing this as it has hit a deadend and there's no motivation of completing it
